sync_check/diff.go

```diff
@@ -0,0 +1,677 @@
+			eq, err := df.checkChunkDataEqual(checkJobs, table)
+			if err != nil {
+				log.Errorf("check chunk data equal failed, error %v", err)
+			}
+			checkResultCh <- eq
+		}()
+	}
+
+	num := 0
+	equal := true
+
+CheckResult:
+	for {
+		select {
+		case eq := <-checkResultCh:
+			num++
+			if !eq {
+				equal = false
+			}
+			if num == df.checkThCount {
+				break CheckResult
+			}
+		}
+	}
+	return equal, nil
+}
+
+func (df *Diff) checkChunkDataEqual(dumpJobs []*util.DumpJob, table *TableCheckCfg) (bool, error) {
+	if len(dumpJobs) == 0 {
+		return true, nil
+	}
+
+	for _, job := range dumpJobs {
+		log.Infof("check table: %s, range: %s", job.Table, job.Where)
+
+		rows1, orderKeyCols, err := getChunkRows(df.db1, df.dbName, table, job.Where, df.useRowID)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		defer rows1.Close()
+
+		rows2, _, err := getChunkRows(df.db2, df.dbName, table, job.Where, df.useRowID)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		defer rows2.Close()
+
+		cols1, err := rows1.Columns()
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		cols2, err := rows2.Columns()
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if len(cols1) != len(cols2) {
+			return false, errors.Trace(err)
+		}
+
+		eq, err := df.compareRows(rows1, rows2, orderKeyCols, table)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if !eq {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (df *Diff) compareRows(rows1, rows2 *sql.Rows, orderKeyCols []*model.ColumnInfo, table *TableCheckCfg) (bool, error) {
+	equal := true
+	rowsData1 := make([]map[string][]byte, 0, 100)
+	rowsData2 := make([]map[string][]byte, 0, 100)
+
+	for rows1.Next() {
+		data1, err := pkgdb.ScanRow(rows1)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		rowsData1 = append(rowsData1, data1)
+	}
+	for rows2.Next() {
+		data2, err := pkgdb.ScanRow(rows2)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		rowsData2 = append(rowsData2, data2)
+	}
+
+	var index1, index2 int
+	for {
+		if index1 == len(rowsData1) {
+			// the all rowsData2's data should be deleted
+			for ; index2 < len(rowsData2); index2++ {
+				sql := generateDML("delete", rowsData2[index2], orderKeyCols, table.Info, table.Schema)
+				log.Infof("[delete] sql: %v", sql)
+				df.wg.Add(1)
+				df.sqlCh <- sql
+				equal = false
+			}
+			break
+		}
+		if index2 == len(rowsData2) {
+			// rosData2 lack some data, should insert them
+			for ; index1 < len(rowsData1); index1++ {
+				sql := generateDML("replace", rowsData1[index1], orderKeyCols, table.Info, table.Schema)
+				log.Infof("[insert] sql: %v", sql)
+				df.wg.Add(1)
+				df.sqlCh <- sql
+				equal = false
+			}
+			break
+		}
+		eq, cmp, err := compareData(rowsData1[index1], rowsData2[index2], orderKeyCols)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if eq {
+			index1++
+			index2++
+			continue
+		}
+		equal = false
+		switch cmp {
```

Comment:
for example:
table1:
+----+------+
| id     | name |
+----+------+
|  1     | a        |
|  2     | b       |
|  4     | e       |
+----+------+

table2:
+----+------+
| id     | name |
+----+------+
|  1     | a         |
|  3     | c         |
|  4     | d        |
+----+------+

id is the primary key, and is also the order by key when select.
1. index1 = 0, index2 = 0
1, a is equal to 1, a
so equal is true, cmp is 0
index1++
index2++

2. index1 = 1, index2 = 1
2, b not equal to 3, c, and the primary key 2 is less than 3
so equal is false, and cmp is -1
need generate insert sql for fix data
index1++

3. index1 = 2, index2 = 1
4, e not equal to 3, c, and the primary key 4 is greater than 3
so equal is false, and cmp is 1
need generate delete sql for fix data
index2++

4. index1 = 2, index2 = 2
4, e not equal to 4, d, and the primary key 4 is equal.
so equal is false, and cmp is 0
need generate update sql for fix data
index1++, index2++